src/explore-education-statistics-common/src/modules/charts/components/utils/generateLegendDataClasses.ts

```diff
@@ -0,0 +1,204 @@
+import { DataClassification } from '@common/modules/charts/types/chart';
+import subtract from '@common/utils/math/subtract';
+import countDecimalPlaces from '@common/utils/number/countDecimalPlaces';
+import formatPretty, {
+  defaultMaxDecimalPlaces,
+} from '@common/utils/number/formatPretty';
+import roundNearest, {
+  roundUpToNearest,
+} from '@common/utils/number/roundNearest';
+import { getScale } from 'color2k';
+import clamp from 'lodash/clamp';
+import orderBy from 'lodash/orderBy';
+import times from 'lodash/times';
+
+export interface LegendDataClass {
+  colour: string;
+  min: string;
+  minRaw: number;
+  max: string;
+  maxRaw: number;
+}
+
+interface Options {
+  colour: string;
+  classification?: DataClassification;
+  decimalPlaces?: number;
+  groups: number;
+  values: number[];
+  unit?: string;
+}
+
+/**
+ * Creates a range of data classes expressing min/max
+ * ranges and a scaled colour for each of those classes.
+ *
+ * Useful for choropleth maps or where data should be
+ * categorised into discrete groups.
+ */
+export default function generateLegendDataClasses({
+  colour,
+  classification = 'EqualIntervals',
+  decimalPlaces: explicitDecimalPlaces,
+  groups,
+  values: initialValues,
+  unit,
+}: Options): LegendDataClass[] {
+  let implicitDecimalPlaces = 0;
+
+  const values = orderBy(initialValues, value => {
+    // Figure out the highest number of decimal places in the values if
+    // no explicit decimal places have been set. We do this whilst ordering
+    // to avoid having to loop over everything again (as an optimisation).
+    if (typeof explicitDecimalPlaces !== 'number' && Number.isFinite(value)) {
+      const dp = countDecimalPlaces(value.toString()) ?? 0;
+
+      if (dp > implicitDecimalPlaces) {
+        implicitDecimalPlaces = dp;
+      }
+    }
+
+    return value;
+  });
+
+  let decimalPlaces =
```

Comment:
defaultMaxDecimalPlaces isn't being respected here, I'm getting 5 decimal places on this branch where on dev it's limited to 2. Also, the colours are a lot duller, is that intended?

![map-pr](https://user-images.githubusercontent.com/81572860/189663844-909a8b61-430b-4e7e-be6a-5a5a0e263b7f.PNG)
![map-dev](https://user-images.githubusercontent.com/81572860/189663852-15ff8d4c-773a-4fb5-b547-69020bddd28b.PNG)

Reply:
Oops, there was a mistake there! Should now respect 2 d.p and there's some tests for it as well.

In terms of the colours - this is mostly intentional. The previous iteration wasn't really scaled properly as it was using some hand-picked values to get a 'nice' palette based on 5 groups. This involved a bunch of manual trial and error so it really wasn't a practical way to generate colours for a user-defined number of groups.

I've currently scaled it from white (the top-end) to the target colour (bottom-end), so this tends towards lighter tones as the range is typically wider. 

Here are some variations where we scale the colour differently:

Target colour + 60% lightness at the top-end:

![image](https://user-images.githubusercontent.com/9917868/189933978-42b8e38c-f478-4b7f-bd9e-dc67d5263f45.png)

Target colour + 50% lightness at the top-end:

![image](https://user-images.githubusercontent.com/9917868/189932145-df84b0df-f3a2-4218-abc4-ea2c8527029a.png)

Target colour + 40% lightness at the top-end:

![image](https://user-images.githubusercontent.com/9917868/189932596-15df91dd-ec31-440b-b890-641d16024768.png)

Not sure if you have any preferences?

Reply:
I'd make it a bit brighter as it's very grey currently. Maybe + 50% lightness?